### ISSUE TITLE
Execd does not parse correctly line protocol with newline characters

### ISSUE THREAD

**Issue by spaghettidba** [influxdata/telegraf#12436]
### Relevant telegraf.conf

```toml
[agent]
  debug = true
  logtarget = "file"
  logfile = '/mnt/c/temp/telegraf/linux/telegraf.log'
# 
# this one works
#
# [[inputs.file]]
#   files = ["sample.txt"]
#   data_format = "influx"

# 
# this one also works
#
# [[inputs.exec]]
#   commands = ['/mnt/c/temp/telegraf/linux/test_exec.sh']
#   timeout = "5s"
#   data_format = "influx"

#
# this one does not
#
[[inputs.execd]]
   command = ['/mnt/c/temp/telegraf/linux/test_execd.sh']
   signal = "none"
   data_format = "influx"

[[outputs.file]]
  files = ["stdout"]
  data_format = "influx"
```

### Logs from Telegraf

```text
2022-12-23T15:52:10Z I! Starting Telegraf 1.24.0
2022-12-23T15:52:10Z I! Available plugins: 222 inputs, 9 aggregators, 26 processors, 20 parsers, 57 outputs
2022-12-23T15:52:10Z I! Loaded inputs: execd
2022-12-23T15:52:10Z I! Loaded aggregators: 
2022-12-23T15:52:10Z I! Loaded processors: 
2022-12-23T15:52:10Z I! Loaded outputs: file
2022-12-23T15:52:10Z I! Tags enabled: host=SQLCLP02
2022-12-23T15:52:10Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"SQLCLP02", Flush Interval:10s
2022-12-23T15:52:10Z D! [agent] Initializing plugins
2022-12-23T15:52:10Z D! [agent] Connecting outputs
2022-12-23T15:52:10Z D! [agent] Attempting connection to [outputs.file]
2022-12-23T15:52:10Z D! [agent] Successfully connected to outputs.file
2022-12-23T15:52:10Z D! [agent] Starting service inputs
2022-12-23T15:52:10Z I! [inputs.execd] Starting process: /mnt/c/temp/telegraf/linux/test_execd.sh []
2022-12-23T15:52:10Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:178: "sqlserver_events_deadlocks,database_name=master,server_instance_name=SRVITBIRDB15\\DBMITBI11,server_principal_name=sa event_sequence=5306688i,session_id=50i,xml_report=\"this does"
2022-12-23T15:52:10Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:10: "not work\" 1671711343642136000"
2022-12-23T15:52:15Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:178: "sqlserver_events_deadlocks,database_name=master,server_instance_name=SRVITBIRDB15\\DBMITBI11,server_principal_name=sa event_sequence=5306688i,session_id=50i,xml_report=\"this does"
2022-12-23T15:52:15Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:10: "not work\" 1671711343642136000"
2022-12-23T15:52:20Z D! [outputs.file] Buffer fullness: 0 / 10000 metrics
2022-12-23T15:52:20Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:178: "sqlserver_events_deadlocks,database_name=master,server_instance_name=SRVITBIRDB15\\DBMITBI11,server_principal_name=sa event_sequence=5306688i,session_id=50i,xml_report=\"this does"
2022-12-23T15:52:20Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:10: "not work\" 1671711343642136000"
2022-12-23T15:52:20Z D! [agent] Stopping service inputs
2022-12-23T15:52:20Z I! [inputs.execd] Process /mnt/c/temp/telegraf/linux/test_execd.sh shut down
2022-12-23T15:52:20Z D! [agent] Input channel closed
2022-12-23T15:52:20Z I! [agent] Hang on, flushing any cached metrics before shutdown
2022-12-23T15:52:20Z D! [outputs.file] Buffer fullness: 0 / 10000 metrics
2022-12-23T15:52:20Z I! [agent] Stopping running outputs
2022-12-23T15:52:20Z D! [agent] Stopped Successfully
```

### System info

Telegraf 1.24.0 Ubuntu

### Steps to reproduce

Use the attached files to reproduce.

1. Run telegraf using the attached telegraf.conf
2. test_execd.sh reads sample.txt every 5 seconds and outputs the contents on stdout
3. telegraf should accept the line protocol data contained in sample.txt through the execd plugin , but it rejects it because it contains a newline char in the field xml_report
4. Observe logs: failed to parse line protocol
5. If you use thisworks.txt (no newline chars) everything works
6. If you use exec instead of execd, the data is accepted (uncomment relevant sections in telegraf.conf)
7. Same for input.file: newline chars are accepted

[repro.zip](https://github.com/influxdata/telegraf/files/10296090/repro.zip)

### Expected behavior

It should accept the data

### Actual behavior

It rejects the data:

2022-12-23T15:52:10Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:178: "sqlserver_events_deadlocks,database_name=master,server_instance_name=SRVITBIRDB15\\DBMITBI11,server_principal_name=sa event_sequence=5306688i,session_id=50i,xml_report=\"this does"
2022-12-23T15:52:10Z E! [inputs.execd] Error in plugin: parse error: metric parse error: expected field at 1:10: "not work\" 1671711343642136000"

### Additional info

Tested and reproduced with all telegraf releases from 1.17 to 1.24.4 both on Windows and Linux

**Comment by spaghettidba** [comment]
Ping @Trovalo :)

**Comment by Trovalo** [comment]
This seems kind of familiar but we need to look into it, I had a similar problem in the past (ref #7406), in that case the issue was 
> It looks like this is a mismatch between what values we allowed in the parser vs the serializer. Using \r was not allowed unless it was followed by \n.

now I had a look at the [Line Protocol Docs](https://docs.influxdata.com/influxdb/v1.8/write_protocols/line_protocol_reference/#line-protocol-syntax) and it seems conflicting... as it says:
> Line protocol accepts the newline character \n and is whitespace-sensitive.
> Note Line protocol does not support the newline character \n in tag values or field values.

Given the issue, seems like the latter is the "true" one.

That said, we need to check what's the actual encoding of the newline character, in my old issue it was `\r\n` and after the fix it worked fine

This means that the following data point must be "encoded" differently
`
##Actual (wrong)
sqlserver_events_deadlocks,database_name=master,server_instance_name=SRVITBIRDB15\DBMITBI11,server_principal_name=sa event_sequence=5306688i,session_id=50i,xml_report="this does
not work" 1671711343642136000
##Desired (correct)
sqlserver_events_deadlocks,database_name=master,server_instance_name=SRVITBIRDB15\DBMITBI11,server_principal_name=sa event_sequence=5306688i,session_id=50i,xml_report="this does \r\n not work" 1671711343642136000
##To be tested (to check which line of the docs is correct....)
sqlserver_events_deadlocks,database_name=master,server_instance_name=SRVITBIRDB15\DBMITBI11,server_principal_name=sa event_sequence=5306688i,session_id=50i,xml_report="this does \n not work" 1671711343642136000
`

Since you are passing line protocol the encoding has to be correct in the source itself, line protocol wise the newline must be an escaped character sequence (as newline means end of the point)

**Comment by Trovalo** [comment]
Update.... 
I've just re-read `6.If you use exec instead of execd, the data is accepted` and this makes no sense.... something has to be wrong with execd

**Comment by spaghettidba** [comment]
More information on this topic.

1. As pointed out by @Trovalo , the docs are not consistent regarding whether newline characters are allowed or not in fields or tags
2. The line protocol documentation does not state anywhere how to escape newline characters properly. It does not even say whether escaping them is required or not.

That said, I tried to circumvent the problem using the json parser with exec and execd and the results are surprising. Execd does not parse json correctly, while exec does. There is something going on with execd that eludes me.

Here is how to repro with the json format, using the attached repro_json.zip:

```toml
[agent]
  debug = true
  logtarget = "file"
  logfile = '/mnt/c/temp/telegraf/linux/telegraf.log'
# 
# this one works
#
# [[inputs.file]]
#   files = ["sample.json"]
#   data_format = "json"

# 
# this one also works
#
# [[inputs.exec]]
#   commands = ['/mnt/c/temp/telegraf/linux/test_exec_json.sh']
#   timeout = "5s"
#   data_format = "json"
#   json_name_key = "measurement"
#   tag_keys = ["server_instance_name","event_sequence"]
#   json_string_fields = ["xml_report","database_name","server_principal_name"]
#   json_time_key = "timestamp"
#   json_time_format = "unix_ns"
#   json_timezone = "UTC"

#
# this one does not
#
[[inputs.execd]]
  command = ['/mnt/c/temp/telegraf/linux/test_execd_json.sh']
  signal = "none"
  data_format = "json"
  json_name_key = "measurement"
  tag_keys = ["server_instance_name","event_sequence"]
  json_string_fields = ["xml_report","database_name","server_principal_name"]
  json_time_key = "timestamp"
  json_time_format = "unix_ns"
  json_timezone = "UTC"


[[outputs.file]]
  files = ["stdout"]
  data_format = "influx"
```

Using exec I get the contents of sample.json converted to line protocol. It is interesting to note that [[outputs.file]] transforms my string with newline to the following line protocol format:

```
sqlserver_events_deadlocks,event_sequence=5306688,host=SQLCLP02,server_instance_name=SRVITBIRDB15\DBMITBI11 database_name="master",server_principal_name="sa",session_id=50,xml_report="this does
not work" 1671711344000000000
```

Please note the newline in the data above. Since [[outputs.file]] uses the influx serializer, I assume that newlines are allowed in fields, unless the influx serializer is also broken.

Using execd I get an error:

```log
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: unexpected end of JSON input
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character ':' after top-level value
2023-01-04T16:14:39Z E! [inputs.execd] Error in plugin: parse error: invalid character '}' looking for beginning of value
```

**Long story short: execd parses input differently, both for influx and json format. This does not appear to be correct.**

[repro_json.zip](https://github.com/influxdata/telegraf/files/10346131/repro_json.zip)

**Comment by powersj** [comment]
> As pointed out by @Trovalo , the docs are not consistent regarding whether newline characters are allowed or not in fields or tags

I believe these are allowed. If you look at the tests for the `influx` parser as well as `exec` and `execd` input plugins, they have cases for `\n` in a field. In fact someone filed an issue, https://github.com/influxdata/telegraf/issues/7461, for what looks like this very issue with execd and newlines in strings. It was fixed and a test added to avoid this in the future, but looking at the set up, it does not parse the incoming data in the same manner.

> Long story short: execd parses input differently, both for influx and json format. This does not appear to be correct.

`inputs.exec` runs your command, sends the output, a byte slice, directly to the parser's parse function. [See here](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/exec/exec.go#L145-L152). As trovalo pointed out, exec does remove the windows return character `\r` or `0x0D` on systems running on Windows. This should not cause an issue in your case as the newline characters would be preserved.

`inputs.execd` has two functions for reading the output, `cmdReadOut` and `cmdReadOutStream`. The latter is used if [using the influx parser](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/execd/execd.go#L49). These two functions make use of `scanner.Scan` and `parser.Next` to read from the resulting `io.Reader` and send data to the parser.

As an aside, keep in mind we have two line protocol parsers: the current default is `influx` and it was a hand-built, in-tree parser. This is not the go forward option. We plan to move to the `influx_upstream` parser by default in a future release. This is the same [upstream parser](https://github.com/influxdata/line-protocol) that is used by InfluxDB Cloud and is the authoritative source. I do not think switching will make a difference in this case due to the different processing methods of `execd`.

Going back to the `scanner.Scan` and `parser.Next`, what I think is happening is both functions are splitting on newlines ([like this example](https://goplay.tools/snippet/y_i48nTlHqc)).

Does that seem to line up with your observations?

**Comment by spaghettidba** [comment]
Makes sense, thanks for your comments. I don't speak go, but I could follow through your comments and they gave me more context to the issue.
Yes, it does line up with my tests.
**So, if I understand correctly, we are dealing with a bug, right?** 

> As trovalo pointed out, **exec does remove the windows return character \r or 0x0D on systems running on Windows**. This should not cause an issue in your case as the newline characters would be preserved.

This doesn't look like a a good idea to me. Keep in mind that telegraf running on windows could receive data with linux newlines and vice-versa. Altering the data without notice could have unexpected consequences.

**Example:** query from SQL Server contains windows newlines, a data point is captured with exec and LF chars are removed. Later, you look up the query by text from the live cache to retrieve plans or runtime stats and you get nothing because the text needs to be matched **exactly**, including whitespaces and non-printable chars. 

This has nothing to do with the issue itself, but I think it is worth investigating.

**Comment by powersj** [comment]
>  So, if I understand correctly, we are dealing with a bug, right?

This is what I am not sure about yet. This behavior has existed since this plugin has been around and I want to ensure I am not going to break existing users if we make any changes here or make a change for behavior that was picked for a very specific reason.

I think from a high-level the fix would be to try passing the entire message to the parse as the other plugins do. Then we would need to re-test this plugin with multiple messages as well as with the external plugins (if they apply here, can't recall off-hand).

Sven and I will chat about this on Wednesday.

**Comment by spaghettidba** [comment]
Thanks, appreciate it 👍

**Comment by powersj** [comment]
After chatting about this it does seem like this should be a config option to change the behavior for this use case, similar to the multiline options with tail plugin. We are concerned that because execd can seemingly send in random amounts of data it might not be good to change the default behavior due to other use-cases, especially for external plugins.

next steps: look at bringing over the multiline parsing functionality and making it common between the two plugins to avoid duplication.

**Comment by spaghettidba** [comment]
Thanks for the feedback. I switched to JSON input, so I can safely send every point in a single line, escaping newlines in a documented way. I could do the same with line protocol, if only escaping newlines was supported.

Thanks again for looking into this!

**Comment by srebhan** [comment]
@spaghettidba I tested using `data_format = "influx_upstream"` and this accepts lines in the form
```text
test field="some line\n another line" 1707149694000000000
```

**Comment by spaghettidba** [comment]
Hey! Glad you got it working. Honestly, 1 year has passed and I can't recall the details of this issue.
IIRC, the central point was whether newlines (**both Unix and windows**) can be escaped and how. The docs are not conclusive about this. 
Your tests seem to be proving \n in a field does not break the parser. Does this also insert a Unix newline in the database? Or does it insert a backlash and an n? Does it work with windows newlines?
At the time, I provided all the test data and scripts to reproduce the issue. Did you run the scripts I provided, combined with your proposed solution?
Thanks for looking into this

**Comment by srebhan** [comment]
@spaghettidba Just tested the line you provided, put it in a file and use the file input. The metric produced will have the newline in the field. For windows this will be CRLF instead of LF so this will end up in the output you use I think.

The issue producing the error is in the `influx`  parser and this probably should be fixed...

**Comment by spaghettidba** [comment]
Well that's phenomenal! Thank you very much!
So the key to make it work is using the influx_upstream dataformat?

**Comment by srebhan** [comment]
@spaghettidba exactly, however, we keep this open until the influx parser is fixed as this is actually an issue...

**Comment by srebhan** [comment]
After discussing this with folks internally, we are going to close this issue as you can use `influx_upstream` .
